typing/ctype.ml

```diff
@@ -2916,14 +2916,34 @@ let is_contractive env p =
 exception Occur
 
 let rec occur_rec env allow_recursive visited ty0 ty =
+  let check_in_children ~visited ty =
+    (* [visited] vs [expanded]: The [visited] set is necessary for correctness,
+       as it ensures that we correctlty handle recursive types and avoid
+       infinite loops. [expanded], however, is an optimization that prevents us
+       from expanding two sibling nodes that are physically equal, as it would
+       be redundant to do so. This prevents being exponential in types like:
+         type 'a t0 = ('a * 'a)
+         type 'a t1 = ('a * 'a) t0
+         type 'a t2 = ('a * 'a) t1
+         ... *)
+    let expanded = TypeHash.create 13 in
```

Comment:
In an ideal world we'd use an immutable map/set here rather than a hashmap, which we could then thread into our recursive calls. This would allow us to successfully skip the second `B` in cases like:
```
  A
 / \
B   C
    |
    B
```
However, the stdlib doesn't offer an easy way to base a map/set off of phys_equal checks. (Note that `TypeSet` compares the `id` field of types whereas `TypeHash` compares physical equality.) We could make something ourselves for this, but it doesn't seem worth the effort right now as we don't have a motivation for improving this case.

Reply:
Why is comparing the `id` field wrong? My understanding is that comparing `id`s should be always equivalent to comparing physical equality.